### PR TITLE
Military spy losses

### PR DESCRIPTION
**Pull request by internetfett**
Kill a percentage of military units with the 'count_as_spy_offense' perk on failed ops.

The percentage of spy units lost is then divided by their DP (currently 4 for both chameleons and master thieves). Normally, losing 1% of 1000 spies would kill 10. In this case, 1% of 5000 chameleons (counts as 1000 spies) would kill 2.5. This can be adjusted as needed with feedback from the GPC.

### CODE REVIEW

**Comment by internetfett** [comment]
Fixes #534

File: src/Helpers/NotificationHelper.php [hunk 1]
```diff
@@ -353,13 +353,20 @@ public function getNotificationMessage(string $category, string $type, array $da
                         throw new \LogicException("Repelled spy op notification for operation key {$data['operationKey']} not yet implemented");
                 }
 
+                $lossString = "";
+                foreach ($data['spiesKilled'] as $key => $value) {
+                    $lossString .= number_format($value) . " of our " . str_plural($key, $value);
```
Comment (WaveHack) [inline src/Helpers/NotificationHelper.php]: I don't think " of our " would make sense here, seeing the context.

I.e. "'Spies from Foo (#1) were discovered in our vaults! We executed 12 of our spies.',

Should probably be "of their".

In addition, the helper function `generate_sentence_from_array()` exists in `src/helpers.php`. Please use that for this

File: src/Helpers/NotificationHelper.php [hunk 2]
```diff
@@ -444,13 +451,20 @@ public function getNotificationMessage(string $category, string $type, array $da
                         throw new \LogicException("Repelled resource theft op notification for operation key {$data['operationKey']} not yet implemented");
                 }
 
+                $lossString = "";
+                foreach ($data['spiesKilled'] as $key => $value) {
+                    $lossString .= number_format($value) . " of our " . str_plural($key, $value);
```
Comment (WaveHack) [inline src/Helpers/NotificationHelper.php]: "of our" => "of their"

File: src/Services/Dominion/Actions/EspionageActionService.php [hunk 3]
```diff
@@ -467,35 +480,48 @@ protected function performResourceTheftOperation(Dominion $dominion, string $ope
...
+                        'spiesKilled' => $unitsKilled,
                     ])
                     ->sendNotifications($target, 'irregular_dominion');
 
-                if ($spiesKilled > 0) {
-                    $message = ("The enemy has prevented our {$operationInfo['name']} attempt and managed to capture " . number_format($spiesKilled) . ' of our spies.');
+                if (!empty($unitsKilled)) {
+                    $lossString = "";
+                    foreach ($unitsKilled as $key => $value) {
+                        $lossString .= number_format($value) . " of our " . str_plural($key, $value);
```
Comment (WaveHack) [inline src/Services/Dominion/Actions/EspionageActionService.php]: Please use `generate_sentence_from_array()`

File: src/Services/Dominion/Actions/EspionageActionService.php [hunk 4]
```diff
@@ -242,35 +242,48 @@ protected function performInfoGatheringOperation(Dominion $dominion, string $ope
...
```
Comment (WaveHack) [inline src/Services/Dominion/Actions/EspionageActionService.php]: Please use `generate_sentence_from_array()`

File: src/Helpers/NotificationHelper.php (hunk 2 above)
Comment (WaveHack) [reply src/Helpers/NotificationHelper.php]: Also please use `generate_sentence_from_array()`

**Comment by internetfett** [comment]
Updated to use generate_sentence_from_array (which is nice btw). I don't think "of our" / "of their" is actually necessary at all here. It didn't exist before my changes.

**Comment by internetfett** [comment]
Also, adjusting/balancing these losses should come from GPC feedback (after merging).

**Comment by WaveHack** [comment]
We can wait with merging until we have feedback from the GPC. Can you in the meantime resolve merge conflicts and merge develop into this branch to update it?

**Comment by internetfett** [comment]
Ready to be merged

**Pull request merged**